kafka-http-connector/main.go

```diff
@@ -0,0 +1,391 @@
+		conn.logger.Error("message received to publish to error topic, but no error topic was set",
+			zap.String("message", err.Error()),
+			zap.String("source", conn.connectorData.SourceName),
+			zap.String("http endpoint", conn.connectorData.HTTPEndpoint),
+		)
+	}
+}
+
+func (conn *kafkaConnector) responseHandler(msg string, headers []sarama.RecordHeader) bool {
+	if len(conn.connectorData.ResponseTopic) > 0 {
+		_, _, err := conn.producer.SendMessage(&sarama.ProducerMessage{
+			Topic:   conn.connectorData.ResponseTopic,
+			Value:   sarama.StringEncoder(msg),
+			Headers: headers,
+		})
+		if err != nil {
+			conn.logger.Warn("failed to publish response body from http request to topic",
+				zap.Error(err),
+				zap.String("topic", conn.connectorData.Topic),
+				zap.String("source", conn.connectorData.SourceName),
+				zap.String("http endpoint", conn.connectorData.HTTPEndpoint))
+			return false
+		}
+	}
+
+	return true
+}
+
+func getProducer(metadata kafkaMetadata) (sarama.SyncProducer, error) {
+	config, err := getConfig(metadata)
+	if err != nil {
+		return nil, err
+	}
+
+	config.Producer.RequiredAcks = sarama.WaitForAll
+	config.Producer.Retry.Max = 10
+	config.Producer.Return.Successes = true
+	producer, err := sarama.NewSyncProducer(metadata.bootstrapServers, config)
+	if err != nil {
+		return nil, err
+	}
+	return producer, nil
+}
+
+func main() {
+	logger, err := zap.NewProduction()
+	if err != nil {
+		log.Fatalf("can't initialize zap logger: %v", err)
+	}
+	defer logger.Sync()
+
+	metadata, err := parseKafkaMetadata(logger)
+	if err != nil {
+		logger.Error("Failed to fetch kafka metadata", zap.Error(err))
+		return
+	}
+
+	connData, err := common.ParseConnectorMetadata()
+	if err != nil {
+		logger.Error("Failed to parse connector meta data", zap.Error(err))
+		return
+	}
+
+	config, err := getConfig(metadata)
+	if err != nil {
+		logger.Error("Failed to create kafka config", zap.Error(err))
+		return
+	}
+
+	producer, err := getProducer(metadata)
+	if err != nil {
+		logger.Error("Failed to create kafka producer", zap.Error(err))
+		return
+	}
+	defer producer.Close()
+
+	conn := kafkaConnector{
+		ready:         make(chan bool),
+		logger:        logger,
+		producer:      producer,
+		connectorData: connData,
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	client, err := sarama.NewConsumerGroup(metadata.bootstrapServers, metadata.consumerGroup, config)
+	if err != nil {
+		logger.Error("Error creating consumer group client", zap.Error(err))
+		return
+	}
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+		for {
+			if err := client.Consume(ctx, []string{connData.Topic}, &conn); err != nil {
+				logger.Error("Error from consumer", zap.Error(err))
+			}
+			// check if context was cancelled, signaling that the consumer should stop
+			if ctx.Err() != nil {
+				return
+			}
+			conn.ready = make(chan bool)
+		}
+	}()
+
+	<-conn.ready // Await till the consumer has been set up
```

Comment:
I've started reviewing this kafka connector code to understand what it does, as part of my investigation to start using it in production. It seems to me that this ready logic is faulty and would never advance past this point. The initial value of `conn.ready` is `make(chan bool)` which correctly blocks until a value is received. But your goroutine loop never sends a value and for some reason always resets `conn.ready` to another `make(chan bool)`.  This means that while the goroutine will Consume properly, this main line will block forever and never set up signal handling and graceful shutdown.

A correct approach would likely be either:

1. A single `close(conn.ready)` *before* the for-loop to indicate that the consumer loop is just about to start
1. Using a `select` to send a `bool` to the channel, with a `default` case that skips all the subsequent sends after the main is no longer receiving
1. a `sync.Once` that can `close(conn.ready)` in the for-loop, and every subsequent call is a no-op

I guess the intent of this ready signal is meant to happen only after the first `client.Consume` call actually returns once?

Reply:
Nevermind, I was wrong. I forked this project to start doing customizations and realized the channel is closed during the `Setup` interface method that is called by the underlying sarama client before going into the blocking consume. Sorry for the noise!

Reply:
Glad that you figured out, let us know if you have any more questions!!